File: typescript-playwright/src/pages/GalleryPage.ts
import type { Page, Locator } from "@playwright/test";
import { BasePage } from "./BasePage.ts";
import { step } from "../utils/decorators.ts";

export class GalleryPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  get searchInput(): Locator {
    return this.page.locator('[data-testid="search-input"]');
  }
  get searchClear(): Locator {
    return this.page.locator('[data-testid="search-clear"]');
  }
  get sidebar(): Locator {
    return this.page.locator('[data-testid="sidebar"]');
  }
  get categoryList(): Locator {
    return this.page.locator('[data-testid="category-list"]');
  }
  get burgerButton(): Locator {
    return this.page.locator('[data-testid="burger-button"]');
  }
  get emptyGallery(): Locator {
    return this.page.locator('[data-testid="empty-gallery"]');
  }
  get photoCardList(): Locator {
    return this.page.locator('[data-testid="photo-card-list"]');
  }
  get loadingSpinner(): Locator {
    return this.page.locator('[data-testid="loading-spinner"]');
  }
  get subcategoryList(): Locator {
    return this.page.locator('[data-testid="subcategory-list"]');
  }

  private categoryDeleteBtn(name: string): Locator {
    return this.categoryItemByName(name).locator(
      '[data-testid="category-delete-btn"]',
    );
  }

  private subcategoryDeleteBtn(name: string): Locator {
    return this.subcategoryItemByName(name).locator(
      '[data-testid="subcategory-delete-btn"]',
    );
  }

  private deletePhotoBtn(photoId: number): Locator {
    return this.page.locator(`[data-testid="photo-card-delete-${photoId}"]`);
  }

  photoImg(photoId: number): Locator {
    return this.page.locator(`[data-testid="photo-img-${photoId}"]`);
  }

  categoryItemByName(name: string): Locator {
    return this.testIdStartsWith("category-item").filter({ hasText: name });
  }

  subcategoryItemByName(name: string): Locator {
    return this.testIdStartsWith("subcategory-item").filter({ hasText: name });
  }

  photoByAlt(altText: string): Locator {
    return this.page.getByAltText(altText);
  }

  @step("Открыть галерею")
  async goto(): Promise<void> {
    await this.page.goto("/");
  }

  @step("Удалить фото")
  async deletePhoto(photoId: number): Promise<void> {
    await this.deletePhotoBtn(photoId).click();
    await this.photoImg(photoId).waitFor({ state: "detached", timeout: 5000 });
  }

  @step("Поиск фото")
  async searchPhoto(searchQuery: string): Promise<void> {
    await this.searchInput.fill(searchQuery);
  }

  @step("Очистить поиск")
  async clearSearch(): Promise<void> {
    await this.searchClear.waitFor({ state: "visible", timeout: 3000 });
    await this.searchClear.click();
  }

  @step("Выбрать категорию по имени")
  async selectCategoryByName(name: string): Promise<void> {
    const category = this.categoryItemByName(name);
    const ariaExpanded = await category.getAttribute("aria-expanded");

    if (ariaExpanded === "false") {
      await category.click();
      await this.subcategoryList.waitFor({ state: "visible", timeout: 5000 });
    } else if (ariaExpanded === null) {
      await category.click();
    }
  }

  @step("Выбрать подкатегорию по имени")
  async selectSubcategoryByName(name: string): Promise<void> {
    await this.subcategoryItemByName(name).click();
  }

  @step("Удалить категорию")
  async deleteCategory(name: string): Promise<void> {
    await this.categoryDeleteBtn(name).click();
  }

  @step("Удалить подкатегорию")
  async deleteSubcategory(name: string): Promise<void> {
    await this.subcategoryDeleteBtn(name).click();
    await this.subcategoryItemByName(name).waitFor({
      state: "detached",
      timeout: 5000,
    });
  }

  @step("Открыть фото")
  async openPhoto(name: string): Promise<void> {
    await this.photoByAlt(name).click();
  }

  @step("Получить список всех заголовков фото")
  async getAllPhotoTitles(): Promise<string[]> {
    return await this.photoCardList
      .locator("img")
      .evaluateAll((imgs) => imgs.map((img) => img.getAttribute("alt") || ""));
  }
}

File: python/src/pages/gallery_page.py
import allure
from playwright.async_api import Locator, Page

from src.pages.base_page import BasePage
from src.utils.constants import STATE_DETACHED, STATE_VISIBLE, TIMEOUT_3S, TIMEOUT_5S


class GalleryPage(BasePage):
    URL_PATH = "/"

    def __init__(self, page: Page):
        super().__init__(page)

    @property
    def search_input(self) -> Locator:
        return self.page.get_by_test_id("search-input")

    @property
    def search_clear_btn(self) -> Locator:
        return self.page.get_by_test_id("search-clear")

    @property
    def sidebar(self) -> Locator:
        return self.page.get_by_test_id("sidebar")

    @property
    def category_list(self) -> Locator:
        return self.page.get_by_test_id("category-list")

    @property
    def burger_btn(self) -> Locator:
        return self.page.get_by_test_id("burger-btn")

    @property
    def empty_state_message(self) -> Locator:
        return self.page.get_by_test_id("empty-state-message")

    @property
    def photo_card_list(self) -> Locator:
        return self.page.get_by_test_id("photo-card-list")

    @property
    def loading_spinner(self) -> Locator:
        return self.page.get_by_test_id("loading-spinner")

    @property
    def subcategory_list(self) -> Locator:
        return self.page.get_by_test_id("subcategory-list")

    def category_item_by_name(self, name: str) -> Locator:
        return self.page.locator('[data-testid^="category-item-"]').filter(has_text=name)

    def subcategory_item_by_name(self, name: str) -> Locator:
        return self.page.locator('[data-testid^="subcategory-item-"]').filter(has_text=name)

    def _category_delete_btn(self, name: str) -> Locator:
        parent = self.category_item_by_name(name)
        return parent.get_by_test_id("category-delete-btn")

    def _subcategory_delete_btn(self, name: str) -> Locator:
        parent = self.subcategory_item_by_name(name)
        return parent.get_by_test_id("subcategory-delete-btn")

    def _delete_photo_btn(self, photo_id: int) -> Locator:
        return self.page.get_by_test_id(f"photo-card-delete-{photo_id}")

    def photo_img(self, photo_id: int) -> Locator:
        return self.page.get_by_test_id(f"photo-img-{photo_id}")

    def photo_by_alt(self, alt_text: str) -> Locator:
        return self.page.get_by_alt_text(alt_text)

    @allure.step("Открыть галерею")
    async def goto(self) -> None:
        await self.page.goto(self.URL_PATH)

    @allure.step("Удалить фото")
    async def delete_photo(self, photo_id: int) -> None:
        await self._delete_photo_btn(photo_id).click()
        await self.photo_img(photo_id).wait_for(state=STATE_DETACHED, timeout=TIMEOUT_5S)

    @allure.step("Поиск фото по тексту: {search_query}")
    async def search_photo(self, search_query: str) -> None:
        await self.search_input.fill(search_query)

    @allure.step("Очистить поиск")
    async def clear_search(self) -> None:
        await self.search_clear_btn.wait_for(state=STATE_VISIBLE, timeout=TIMEOUT_3S)
        await self.search_clear_btn.click()

    @allure.step("Выбрать категорию по имени: {name}")
    async def select_category_by_name(self, name: str) -> None:
        category = self.category_item_by_name(name)
        aria_expanded = await category.get_attribute("aria-expanded")

        if aria_expanded == "false":
            await category.click()
            await self.subcategory_list.wait_for(state=STATE_VISIBLE, timeout=TIMEOUT_5S)
        elif not aria_expanded:
            await category.click()

    @allure.step("Выбрать подкатегорию по имени: {name}")
    async def select_subcategory_by_name(self, name: str) -> None:
        await self.subcategory_item_by_name(name).click()

    @allure.step("Удалить категорию: {name}")
    async def delete_category(self, name: str) -> None:
        await self._category_delete_btn(name).click()

    @allure.step("Удалить подкатегорию: {name}")
    async def delete_subcategory(self, name: str) -> None:
        await self._subcategory_delete_btn(name).click()

    @allure.step("Открыть фото: {alt_text}")
    async def open_photo_by_alt(self, alt_text: str) -> None:
        await self.page.get_by_alt_text(alt_text).click()
